test-app/src/main/java/org/readium/r2/testapp/reader/VisualReaderFragment.kt

```diff
@@ -51,49 +75,205 @@ abstract class VisualReaderFragment : BaseReaderFragment(), VisualNavigator.List
         return binding.root
     }
 
+    /**
+     * When true, the user won't be able to interact with the navigator.
+     */
+    private var disableTouches by mutableStateOf(false)
+
     override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
         super.onViewCreated(view, savedInstanceState)
 
         navigatorFragment = navigator as Fragment
 
-        val viewScope = viewLifecycleOwner.lifecycleScope
+        setupObservers()
+
+        childFragmentManager.addOnBackStackChangedListener {
+            updateSystemUiVisibility()
+        }
+        binding.fragmentReaderContainer.setOnApplyWindowInsetsListener { container, insets ->
+            updateSystemUiPadding(container, insets)
+            insets
+        }
+
+        binding.overlay.setContent {
+            if (disableTouches) {
+                // Add an invisible box on top of the navigator to intercept touch gestures.
+                Box(
+                    modifier = Modifier
+                        .fillMaxSize()
+                        .pointerInput(Unit) {
+                            detectTapGestures {
+                                requireActivity().toggleSystemUi()
+                            }
+                        }
+                )
+            }
 
-        navigator.currentLocator
-            .onEach { model.saveProgression(it) }
-            .launchIn(viewScope)
+            Box(
+                modifier = Modifier
+                    .fillMaxSize()
+                    .systemBarsPadding(),
+                content = { Overlay() }
+            )
+        }
+    }
 
+    @Composable
+    private fun BoxScope.Overlay() {
+        model.tts?.let { tts ->
+            TtsControls(
+                model = tts,
+                modifier = Modifier
+                    .align(Alignment.BottomCenter)
+                    .padding(8.dp)
+            )
+        }
+    }
+
+    private fun setupObservers() {
+        viewLifecycleOwner.lifecycleScope.launch {
+            viewLifecycleOwner.repeatOnLifecycle(Lifecycle.State.STARTED) {
+                navigator.currentLocator
+                    .onEach { model.saveProgression(it) }
+                    .launchIn(this)
+
+                setupHighlights(this)
+                setupSearch(this)
+                setupTts(this)
```

Comment:
I split the setup to make it clearer to people wanting to integrate only one of these features.